### ISSUE TITLE
copy attachments between issus

### ISSUE THREAD

**Issue by bitcoder** [sumoheavy/jira-ruby#295]
Hi guys,
I'm trying to figure out a clean way of copying attachments between issues. Unfortunately this seems to be a bit tricky though because attachments have to be uploaded from local filesystem to the remote Jira server. Can the current API be somehow extended in order to: whenever uploading a file also allow it to read it  from the current server as the source?

**Comment by puneet18190** [comment]
@bitcoder I am having same problem to get attachement from one issue and upload it to another issue. Do you find any solutions? Do you know any jira rest api for that?

**Comment by bitcoder** [comment]
@puneet18190, I don't think so; I was reviewing my code that was using this API and I hadn't implemented this part back then; if I remember it correctly, you have to download the attachments to local files and then upload them manually; in other words, you have to manage this by yourself.
I implemented cloning manually for example, as  Jira API does not provide endpoints for this (https://docs.atlassian.com/software/jira/docs/api/REST/8.6.0/?_ga=2.39731171.1970630417.1578137684-1302790130.1547822957)

**Comment by puneet18190** [comment]
@bitcoder how to download attachments locally? In response I got attachment link but its private and I can't fetch the file object using attachment link so how to download/fetch file object ?

**Comment by marlinpierce** [comment]
The cleanest way to do this is if Jira provided an API which handles this on the server without having to download and then upload the attachment.  Jira does not seem to provide that.

In the `jira-ruby 3.0.0` version, there is a new convenience method to download the attachment contents.  @puneet18190 this may answer your question, but I don't know if you had a problem with privileges.

The upload uses a multipart form upload.  Such an HTTP request can be streamed.  Instead of building the entire HTTP request in memory then sending it to the network, it can write the HTTP headers and body, then for the attachment, an IO object can be used to stream the contents, appending and writing the HTTP request as it goes.

The `Attachment#download_file` method will stream the download.

Since Jira does not provide a server side implementation the best you can do is download and upload, streaming both, reading a chunk of content at a time then writing it for the upload without ever saving the file to the file system or reading the entire file in memory at any point.  The contents are held in memory a chunk at a time.

This avoids the time needed for the file system IO.  The alternation between reading and writing the two streams could provide some parallelism, which if copying between different Jira instances could be a big benefit.  With the same Jira server, there may be contention over the same network interface an route.  However, maybe the read and write directions won't contend so much.

Maybe the `jira-ruby` gem could implement a connivance method to do this.
